`-Zdump-mir` introduces cycle in static containing static reference

```rust
// src/lib.rs

pub struct Thing {
    pub next: &'static Thing,
}

pub static THING: Thing = Thing { next: &THING };
```

```console
$ cargo clean
$ RUSTFLAGS=-Zdump-mir= cargo build

error[E0391]: cycle detected when evaluating initializer of static `THING`
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
  |
note: ...which requires caching mir of `THING` for CTFE...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires elaborating drops for `THING`...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `THING`...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `THING`...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const checking `THING`...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires building MIR for `THING`...
 --> src/lib.rs:7:1
  |
7 | pub static THING: Thing = Thing { next: &THING };
  | ^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires evaluating initializer of static `THING`, completing the cycle
  = note: cycle used when running analysis passes on this crate
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

For more information about this error, try `rustc --explain E0391`.
error: could not compile `repro` (lib) due to 1 previous error
```

The error does not occur if you build without `-Zdump-mir` first.

```console
$ cargo clean
$ cargo build
$ RUSTFLAGS=-Zdump-mir= cargo build
```

### Meta

```
rustc 1.85.0-nightly (21fe748be 2024-12-11)
binary: rustc
commit-hash: 21fe748be15271ea5804e0507cd699b675efe038
commit-date: 2024-12-11
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.5
```

--- a/compiler/rustc_middle/src/mir/pretty.rs
+++ b/compiler/rustc_middle/src/mir/pretty.rs
@@ -1555,16 +1555,22 @@ pub fn write_allocations<'tcx>(
                 write!(w, " (vtable: impl {dyn_ty} for {ty})")?
             }
             Some(GlobalAlloc::Static(did)) if !tcx.is_foreign_item(did) => {
-                match tcx.eval_static_initializer(did) {
-                    Ok(alloc) => {
-                        write!(w, " (static: {}, ", tcx.def_path_str(did))?;
-                        write_allocation_track_relocs(w, alloc)?;
+                write!(w, " (static: {}", tcx.def_path_str(did))?;
+                if body.phase <= MirPhase::Runtime(RuntimePhase::PostCleanup)
+                    && tcx.hir().body_const_context(body.source.def_id()).is_some()
+                {
+                    // Statics may be cyclic and evaluating them too early
+                    // in the MIR pipeline may cause cycle errors even though
+                    // normal compilation is fine.
+                    write!(w, ")")?;
+                } else {
+                    match tcx.eval_static_initializer(did) {
+                        Ok(alloc) => {
+                            write!(w, ", ")?;
+                            write_allocation_track_relocs(w, alloc)?;
+                        }
+                        Err(_) => write!(w, ", error during initializer evaluation)")?,
                     }
-                    Err(_) => write!(
-                        w,
-                        " (static: {}, error during initializer evaluation)",
-                        tcx.def_path_str(did)
-                    )?,
                 }
             }
             Some(GlobalAlloc::Static(did)) => {

--- a/tests/mir-opt/building/dump_mir_cycle.rs
+++ b/tests/mir-opt/building/dump_mir_cycle.rs
@@ -1,0 +1,19 @@
+#[derive(Debug)]
+pub struct Thing {
+    pub next: &'static Thing,
+}
+
+pub static THING: Thing = Thing { next: &THING };
+// CHECK: alloc{{.+}} (static: THING)
+
+const fn thing() -> &'static Thing {
+    &MUTUALLY_RECURSIVE
+}
+
+pub static MUTUALLY_RECURSIVE: Thing = Thing { next: thing() };
+// CHECK: alloc{{.+}} (static: MUTUALLY_RECURSIVE)
+
+fn main() {
+    // Generate optimized MIR for the const fn, too.
+    thing();
+}

--- a/tests/mir-opt/const_promotion_extern_static.BAR-promoted[0].SimplifyCfg-pre-optimizations.after.mir
+++ b/tests/mir-opt/const_promotion_extern_static.BAR-promoted[0].SimplifyCfg-pre-optimizations.after.mir
@@ -15,6 +15,4 @@ const BAR::promoted[0]: &[&i32; 1] = {
     }
 }
 
-ALLOC0 (static: Y, size: 4, align: 4) {
-    2a 00 00 00                                     │ *...
-}
+ALLOC0 (static: Y)

--- a/tests/mir-opt/const_promotion_extern_static.BAR.PromoteTemps.diff
+++ b/tests/mir-opt/const_promotion_extern_static.BAR.PromoteTemps.diff
@@ -38,9 +38,7 @@
       bb2 (cleanup): {
           resume;
       }
-- }
-- 
-- ALLOC0 (static: Y, size: 4, align: 4) {
--     2a 00 00 00                                     │ *...
   }
+- 
+- ALLOC0 (static: Y)
   

